File: Problems/No.732_My_Calendar_III/my_calendar_III.py
class MyCalendarThree(object):

    def __init__(self):
        self.calendar = {}
        self.largest_booking = 0

    def book(self, startTime, endTime):
        """
        :type startTime: int
        :type endTime: int
        :rtype: int
        """
        self.calendar[startTime] = self.calendar.get(startTime, 0) + 1
        self.calendar[endTime] = self.calendar.get(endTime, 0) - 1

        partial_sum, cal_booking = 0, 0
        for [time, SoE] in sorted(self.calendar.items()):
            if (time <= endTime):
                partial_sum += SoE
                cal_booking = max(cal_booking, partial_sum)
            else:
                break
        self.largest_booking = max(self.largest_booking, cal_booking)

        return self.largest_booking
        


# Your MyCalendarThree object will be instantiated and called as such:
# obj = MyCalendarThree()
# param_1 = obj.book(startTime,endTime)

File: Problems/No.732_My_Calendar_III/my_calendar_III.cpp
class MyCalendarThree {

private:
    std::map<int, int> Calendar;
    int largest_booking = 0;

public:
    MyCalendarThree() {}
    
    int book(int startTime, int endTime) {
        
        Calendar[startTime]++;
        Calendar[endTime]--;
        
        int sum = 0;
        int cal_booking = 0;

        for (auto &[time, SoE]: Calendar) {

            if (time <= endTime) {
                
                sum += SoE;
                cal_booking = std::max(cal_booking, sum);

            }
            else break;

        }
        largest_booking = std::max(largest_booking, cal_booking);

        return largest_booking;

    }
};

/**
 * Your MyCalendarThree object will be instantiated and called as such:
 * MyCalendarThree* obj = new MyCalendarThree();
 * int param_1 = obj->book(startTime,endTime);
 */
